provider/apple.go

```diff
@@ -0,0 +1,529 @@
+// See more: https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api
+// and https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/json"
+	"golang.org/x/oauth2"
+
+	"encoding/pem"
+	"fmt"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/go-pkgz/auth/logger"
+	"github.com/go-pkgz/auth/token"
+	"github.com/go-pkgz/rest"
+	"github.com/pkg/errors"
+	"io/ioutil"
+
+	"net/http"
+	"net/url"
+	"os"
+
+	"strings"
+	"time"
+)
+
+const (
+	// appleAuthURL is the base authentication URL for sign in with Apple ID and fetch request code for user validation request.
+	appleAuthURL = "https://appleid.apple.com/auth/authorize"
+
+	// appleTokenURL is the endpoint for verifying tokens and get user unique ID and E-mail
+	appleTokenURL = "https://appleid.apple.com/auth/token" // #nosec
+
+	// appleRequestContentType is the valid type which apple REST API accept only
+	appleRequestContentType = "application/x-www-form-urlencoded"
+
+	// UserAgent required to every request to Apple REST API
+	defaultUserAgent = "github.com/go-pkgz/auth"
+
+	// AcceptJSONHeader is the content to accept from response
+	AcceptJSONHeader = "application/json"
+)
+
+// AppleVerificationResponse is based on https://developer.apple.com/documentation/signinwithapplerestapi/tokenresponse
+type AppleVerificationResponse struct {
+	// A token used to access allowed user data, but now not implemented public interface for it.
+	AccessToken string `json:"access_token"`
+
+	// Access token type, always equal the "bearer".
+	TokenType string `json:"token_type"`
+
+	// Access token expires time in seconds. Always equal 3600 seconds (1 hour)
+	ExpiresIn int `json:"expires_in"`
+
+	// The refresh token used to regenerate new access tokens.
+	RefreshToken string `json:"refresh_token"`
+
+	// Main JSON Web Token that contains the user’s identity information.
+	IDToken string `json:"id_token"`
+
+	// Used to capture any error returned in response. Always check error for empty
+	Error string `json:"error"`
+}
+
+// AppleConfig is the main oauth2 required parameters for "Sign in with Apple"
+type AppleConfig struct {
+	ClientID           string   // the identifier Services ID for your app created in Apple developer account.
+	TeamID             string   // developer Team ID (10 characters), required for create JWT. It available, after signed in at developer account, by link: https://developer.apple.com/account/#/membership
+	KeyID              string   // private key ID  assigned to private key obtain in Apple developer account
+	ClientSecretExpire int64    // time in seconds when client secret will expired (default: 30 minutes)
+	Scopes             []string // for apple provider allow only "email" and "name" scope values
```

Comment:
I would suggest droping this one and enforce the lowest scope possible (name?), like we do with other providers. We also could drop user email  upstream

Reply:
Ok.  I will change it.